presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PruneJoinColumns.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Pattern;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.sql.planner.iterative.rule.Util.pruneInputs;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+/**
+ * Non-cross joins support output symbol selection, so absorb any project-off into the node.
```

Comment:
> There needs to be two rules for non-cross join, one that looks for a project over a join, and one that looks for a join, because the join can itself project off columns. Currently, the first rule doesn't need to include the logic to find which join inputs are needed, so that logic can be included inline in the second rule. If the first rule also prunes the join inputs, then the logic would need to be factored out.

This is clear to me, and you already explained that me a couple of times ;)

> I think a separate rule for cross join is cleaner

That is your opinion 

> and matches a future in which NestedLoopJoin is a different plan node than HashJoin, as opposed the current situation where the two operators share a plan node that doesn't actually fit.

IMO NestedLoopJoin, HashJoin, MergeJoin any other join are physical plan nodes, while JoinNode is logical. I would rather expect to have your rules to work on logical nodes. In such case nothing should change. 

Anyway I see you have a strong opinion about this. So let it be as you want. I will create a PR in which I will check my idea and give it to you for a review. Not because I want so much to persuade you, but to just give it a try ;)